Add SRTM dataset

jenkspt opened the pull request:
This is a first attempt at adding the SRTM dataset

Some notes on the implementation:

- The code uses a mirror of SRTM data from [https://srtm.csi.cgiar.org/wp-content/uploads/files/srtm_5x5/TIFF/](https://srtm.csi.cgiar.org/wp-content/uploads/files/srtm_5x5/TIFF/)

- The SRTM dataset is different from the currently implemented datasets:  
    - It has 5x5 degree tiles
    - It's around 400GB (uncompressed)

To handle these constraints, `getraster` takes a bounds argument for selecting the dataset region to download. When called with the bounds argument, this method also returns a matrix of strings with the appropriate tile paths.

TODO:
Need to propose changes so that `Raster(SRTM, bounds, ...)` from `Rasters.jl` can be implemented

jenkspt committed:
add srtm

jenkspt committed:
add srtm tests

jenkspt commented:
Is there a reason the CI is running with Julia 1.5? is it the `[raster_path1;;]` syntax that's causing the error?

rafaqz commented:
Thanks for the PR! This looks great. 

The reason for testing on 1.5 and any errors is simply that I'm spread too thin over other packages and havent had time to work on this as much as I would like to. 

Github actions previously didn't recognise just "1" so you needed the specific version. But it does now, so we should use that and always test on the latest version.

jenkspt commented:
I believe 1.6.5 is now the long term release. From the downloads page [https://julialang.org/downloads/](https://julialang.org/downloads/)

jenkspt commented:
Should `rasterpath` and `rastername` also have methods with the `bounds` argument?

rafaqz commented:
Yep, and `tile_indices` also has to be a keyword arg everywhere. 

The second argument to getraster is always the layer name/s. If there are no layers, there probably should be no second argument.

Otherwise multiple tiles means a RasterStack

rafaqz commented on src/srtm/srtm.jl:
```diff
@@ -0,0 +1,59 @@
+zippath(T::Type{SRTM}, tile_index) = joinpath(rasterpath(T), "zips", zipname(T, tile_index))
+
+
+function getraster(T::Type{SRTM}, tile_index)
```
The second argument to `getraster` is always `layer`. So we should do this with keywords. You could have 
`bounds=nothing, tile_index=wgs84_to_tile_index(bounds)` or something so you can specify either the bounds or the tile.

rafaqz commented on src/srtm/srtm.jl:
```diff
@@ -0,0 +1,59 @@
+end
+
+
+function getraster(T::Type{SRTM}, bounds::NTuple{4,Real})
```
`bounds` should be a keyword in the above getraster method. The problem with mixing them like this is other packages (like Rasters.jl) will think these are four layer numbers for a `RasterStack`, which would be the case for e.g. BioClim datasets. If we can stay consistent we don't have to special-case every dataset and duplicate all the logic.

I may not have got the structure right or general enough, but lets try to do things as consistently as possible and make consistent changes if it turns out we really need to do that.

Also I guess we have to special-case this anyway for stitching the arrays - but the meaning of the arguments being consistent is also a good thing. Other datasets might  (do) have both layers and `bounds`, and this will allow for that

rafaqz commented on src/srtm/srtm.jl:
```diff
@@ -0,0 +1,59 @@
+# SRTM Mirror with 5x5 degree tiles
+const SRTM_URI = URI(scheme = "https", host = "srtm.csi.cgiar.org", path = "/wp-content/uploads/files/srtm_5x5/TIFF")
+
+resolutions(::Type{SRTM}) = ("30m",)
```
If there is only one resolution we probably dont need this. (although maybe its useful to have those in the API? I dont know)

rafaqz commented:
We can test on `1.6` or just `1`, I dont mind. I havent tested multiple versions because CI downloads a lot of files.

jenkspt committed:
add keyword arguments

jenkspt commented:
Ok so I removed the `resolution` and `defres` methods, and added `bounds` and `tile_index` keyword arguments to all of the relevant functions.

Tests are passing locally for me with Julia 1.7.1 -- not exactly sure why the CI is failing.

jenkspt committed:
use eval macro and remove dead code

rafaqz commented on src/srtm/srtm.jl:
```diff
@@ -0,0 +1,79 @@
+        $_op(T::Type{SRTM}, bounds::NTuple{4,Real}) = $_op(T, bounds_to_tile_indices(T, bounds))
+
+        # Public function definition with key-word arguments
+        function $op(T::Type{SRTM}; bounds = nothing, tile_index = nothing)
```
The style here is no spaces around keywords (see BlueStyle), so `bounds=nothing` etc

rafaqz commented on src/srtm/srtm.jl:
```diff
@@ -0,0 +1,79 @@
+                :op === :getraster || return joinpath(rasterpath(), "SRTM")
+                throw(ArgumentError("One of `bounds` or `tile_index` kwarg must be specified"))
+            elseif (bounds !== nothing) & (tile_index !== nothing)
+                throw(ArgumentError("Only on of `bounds` or `tile_index` should be specified. " *
```
```suggestion
                throw(ArgumentError("Only one of `bounds` or `tile_index` should be specified. " *
```

rafaqz commented on src/srtm/srtm.jl:
```diff
@@ -0,0 +1,79 @@
+
+        # Public function definition with key-word arguments
+        function $op(T::Type{SRTM}; bounds = nothing, tile_index = nothing)
+            if (bounds === nothing) & (tile_index === nothing)
```
`isnothing` might read nicer since we use this so much

jenkspt replied:
done

rafaqz reviewed (commented):
LGTM

jenkspt committed:
add BluStyle kwarg spacing + isnothing

jenkspt committed:
Update src/srtm/srtm.jl

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

rafaqz commented:
Have you missed pushing some updates to your tests? 

https://github.com/EcoJulia/RasterDataSources.jl/runs/4833591681?check_suite_focus=true#step:5:1446

rafaqz commented on test/srtm.jl:
```diff
@@ -0,0 +1,26 @@
+    zip_url = URI(scheme = "https", host = "srtm.csi.cgiar.org", path = "/wp-content/uploads/files/srtm_5x5/TIFF/srtm_01_02.zip")
+    tile_index1 = CartesianIndex(2, 1)   # [y, x] order
+    tile_index2 = CartesianIndex(2, 2)   # [y, x] order
+    @test zipurl(SRTM, tile_index = tile_index1) == zip_url
```
always use semi colons before keyword arguments `;`

jenkspt replied:
`;` only after positional arguments? Or do you want `URI(; scheme = "https", host = ...)` as well?

rafaqz replied:
You can do that too if you want, probably BlueStyle does. It just screams "keywords here" at you louder than a comma does

rafaqz reviewed (approved):
This is looking good, we just need to get that test passing and I'll merge

jenkspt committed:
fix `_zipname`

jenkspt committed:
add `;` to kwarg fn calls

rafaqz commented:
Thanks! Turns out I need to use this pretty much straight away in my own work, so looking forward to testing it out.
